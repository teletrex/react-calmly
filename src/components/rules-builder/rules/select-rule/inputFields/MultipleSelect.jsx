/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useContext, useMemo } from 'react';
import PropTypes from 'prop-types';
import filter from 'lodash/filter';
import isEqual from 'lodash/isEqual';
import map from 'lodash/map';
import  {useTranslation}  from '../../../../../translation';
import { MultiSelect} from '@carbon/react';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { PROPTYPES } from '../../constants';
import { getPrefixedClasses } from '../../helpers';

const MultipleSelect = ({ id, isReadOnly, onValueChange, options, placeholder, value }) => {
  const { t } = useTranslation();
  const { isLight } = useContext(RulesBuilderContext);

  const items = useMemo(() => map(options, ({ label, value }) => ({ id: value, text: label })), [
    options,
  ]);

  const selectedItems = filter(items, item => value.includes(item.id));

  const handleOnChange = ({ selectedItems: newItems }) => {
    const formattedItems = map(newItems, item => item.id);
    if (!isEqual(formattedItems, value)) {
      onValueChange(formattedItems);
    }
  };

  if (isReadOnly) {
    return (
      <span className={getPrefixedClasses(['--rule-value-readonly', '--text-space-left'])}>
        {map(selectedItems, item => item.text).join(', ')}
      </span>
    );
  }

  return (
    <MultiSelect
      id={id}
      items={items}
      label={placeholder || t('Choose an option')}
      light={isLight}
      onChange={handleOnChange}
      selectedItems={selectedItems}
      withSearch
    />
  );
};

MultipleSelect.propTypes = {
  id: PropTypes.string.isRequired,
  isReadOnly: PropTypes.bool,
  onValueChange: PropTypes.func.isRequired,
  options: PROPTYPES.SELECT_OPTIONS,
  placeholder: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.string),
};

MultipleSelect.defaultProps = {
  isReadOnly: false,
  options: [],
  placeholder: '',
  value: [],
};

export default MultipleSelect;
