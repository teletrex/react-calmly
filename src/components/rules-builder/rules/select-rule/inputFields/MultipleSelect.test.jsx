/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import MultipleSelect from './MultipleSelect';

const options = [
  { label: 'Option 1', value: 'OPTION_1' },
  { label: 'Option 2', value: 'OPTION_2' },
  { label: 'Option 3', value: 'OPTION_3' },
];
const initialValue = [options[2].value];

const defaultProps = {
  id: 'TestSelectRuleElement',
  isReadOnly: false,
  options,
  onValueChange: jest.fn(),
};

describe('MultipleSelect', () => {
  beforeEach(() => {
    defaultProps.onValueChange.mockReset();
  });

  it('should render correctly', () => {
    const { container } = render(<MultipleSelect {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a provided value', () => {
    const { container } = render(<MultipleSelect {...defaultProps} value={initialValue} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for read-only', () => {
    const { container, getByText } = render(
      <MultipleSelect {...defaultProps} isReadOnly value={initialValue} />
    );
    expect(container).toMatchSnapshot();
    expect(getByText(options[2].label)).toBeInTheDocument();
  });

  it('should set the rule value when a value is selected', () => {
    const { getAllByRole, getByText } = render(<MultipleSelect {...defaultProps} />);
    const triggerButton = getAllByRole('button')[1];

    // expand the value dropdown:
    fireEvent.click(triggerButton);

    // select a new value:
    fireEvent.click(getByText(options[0].label));
    expect(defaultProps.onValueChange).toHaveBeenCalledWith([options[0].value]);
  });

  it('should add newly selected values to the exiting ones', () => {
    const { getAllByRole, getByText } = render(
      <MultipleSelect {...defaultProps} value={initialValue} />
    );
    const triggerButton = getAllByRole('button')[1];

    // expand the value dropdown:
    fireEvent.click(triggerButton);

    // select a new value:
    fireEvent.click(getByText(options[0].label));
    expect(defaultProps.onValueChange).toHaveBeenCalledWith([options[2].value, options[0].value]);
  });

  it('should properly deselect a value', () => {
    const { getAllByRole, getByText } = render(
      <MultipleSelect {...defaultProps} value={initialValue} />
    );
    const triggerButton = getAllByRole('button')[1];

    // expand the value dropdown:
    fireEvent.click(triggerButton);

    // deselect a value:
    fireEvent.click(getByText(options[2].label));
    expect(defaultProps.onValueChange).toHaveBeenCalledWith([]);
  });

  it('should display the default placeholder', () => {
    const { getByText } = render(<MultipleSelect {...defaultProps} />);

    expect(getByText('Choose an option')).toBeInTheDocument();
  });

  it('should display the provided placeholder', () => {
    const { getByText } = render(<MultipleSelect {...defaultProps} placeholder="Test dropdown" />);

    expect(getByText('Test dropdown')).toBeInTheDocument();
  });
});
