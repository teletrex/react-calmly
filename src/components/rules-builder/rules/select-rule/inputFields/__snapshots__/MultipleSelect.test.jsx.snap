// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleSelect should render correctly 1`] = `
<div>
  <div
    class="bx--multi-select__wrapper bx--list-box__wrapper"
  >
    <div
      class="bx--multi-select bx--list-box"
      id="TestSelectRuleElement"
      role="button"
      tabindex="0"
    >
      <button
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-0-label downshift-0-toggle-button"
        class="bx--list-box__field"
        id="downshift-0-toggle-button"
        type="button"
      >
        <span
          class="bx--list-box__label"
          id="multiselect-field-label-multi_select_id_2"
        >
          Choose an option
        </span>
        <div
          class="bx--list-box__menu-icon"
        >
          <svg
            aria-hidden="true"
            focusable="false"
            height="16"
            name="chevron--down"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
            />
            <title />
          </svg>
        </div>
      </button>
      <div
        aria-labelledby="downshift-0-label"
        aria-multiselectable="true"
        class="bx--list-box__menu bx--list-box__close"
        id="downshift-0-menu__menu"
        role="listbox"
        tabindex="-1"
      >
        <div
          aria-selected="false"
          class="bx--dropdown__search-wrapper bx--list-box__menu-item"
          id="downshift-0-item-0"
          role="option"
        >
          <div
            class="bx--list-box__menu-item__option"
          >
            <span
              class="bx--list-box__menu-item__option__text"
            >
              <div
                aria-label="Search..."
                class="bx--search bx--search--sm"
                role="search"
              >
                <svg
                  aria-label="Search..."
                  class="bx--search-magnifier"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  style="will-change: transform;"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 14.3L10.7 10c1.9-2.3 1.6-5.8-.7-7.7S4.2.7 2.3 3 .7 8.8 3 10.7c2 1.7 5 1.7 7 0l4.3 4.3.7-.7zM2 6.5C2 4 4 2 6.5 2S11 4 11 6.5 9 11 6.5 11 2 9 2 6.5z"
                  />
                </svg>
                <label
                  class="bx--label"
                  for="search__input__id_1"
                  id="search__input__id_1-label"
                >
                  Search...
                </label>
                <input
                  aria-label="dropdown-TestSelectRuleElement-search"
                  class="bx--search-input"
                  i18n="[object Object]"
                  id="search__input__id_1"
                  placeholder="Search..."
                  title="Search..."
                  type="text"
                  value=""
                />
                <button
                  aria-label="Clear search input"
                  class="bx--search-close bx--search-close--hidden"
                  title="Clear input"
                  type="button"
                >
                  <svg
                    aria-label="Clear search input"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    style="will-change: transform;"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12 4.7l-.7-.7L8 7.3 4.7 4l-.7.7L7.3 8 4 11.3l.7.7L8 8.7l3.3 3.3.7-.7L8.7 8z"
                    />
                  </svg>
                </button>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultipleSelect should render correctly for read-only 1`] = `
<div>
  <span
    class="bx--rule-value-readonly bx--text-space-left"
  >
    Option 3
  </span>
</div>
`;

exports[`MultipleSelect should render correctly with a provided value 1`] = `
<div>
  <div
    class="bx--multi-select__wrapper bx--list-box__wrapper"
  >
    <div
      class="bx--multi-select bx--multi-select--selected bx--list-box"
      id="TestSelectRuleElement"
      role="button"
      tabindex="0"
    >
      <button
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="downshift-3-label downshift-3-toggle-button"
        class="bx--list-box__field"
        id="downshift-3-toggle-button"
        type="button"
      >
        <div
          aria-label="Clear Selection"
          class="bx--list-box__selection bx--tag--filter bx--list-box__selection--multi"
          role="button"
          tabindex="0"
        >
          1
          <svg
            aria-hidden="true"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 4.7l-.7-.7L8 7.3 4.7 4l-.7.7L7.3 8 4 11.3l.7.7L8 8.7l3.3 3.3.7-.7L8.7 8z"
            />
          </svg>
        </div>
        <span
          class="bx--list-box__label"
          id="multiselect-field-label-multi_select_id_5"
        >
          Choose an option
        </span>
        <div
          class="bx--list-box__menu-icon"
        >
          <svg
            aria-hidden="true"
            focusable="false"
            height="16"
            name="chevron--down"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
            />
            <title />
          </svg>
        </div>
      </button>
      <div
        aria-labelledby="downshift-3-label"
        aria-multiselectable="true"
        class="bx--list-box__menu bx--list-box__close"
        id="downshift-3-menu__menu"
        role="listbox"
        tabindex="-1"
      >
        <div
          aria-selected="false"
          class="bx--dropdown__search-wrapper bx--list-box__menu-item"
          id="downshift-3-item-0"
          role="option"
        >
          <div
            class="bx--list-box__menu-item__option"
          >
            <span
              class="bx--list-box__menu-item__option__text"
            >
              <div
                aria-label="Search..."
                class="bx--search bx--search--sm"
                role="search"
              >
                <svg
                  aria-label="Search..."
                  class="bx--search-magnifier"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  style="will-change: transform;"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 14.3L10.7 10c1.9-2.3 1.6-5.8-.7-7.7S4.2.7 2.3 3 .7 8.8 3 10.7c2 1.7 5 1.7 7 0l4.3 4.3.7-.7zM2 6.5C2 4 4 2 6.5 2S11 4 11 6.5 9 11 6.5 11 2 9 2 6.5z"
                  />
                </svg>
                <label
                  class="bx--label"
                  for="search__input__id_1"
                  id="search__input__id_1-label"
                >
                  Search...
                </label>
                <input
                  aria-label="dropdown-TestSelectRuleElement-search"
                  class="bx--search-input"
                  i18n="[object Object]"
                  id="search__input__id_1"
                  placeholder="Search..."
                  title="Search..."
                  type="text"
                  value=""
                />
                <button
                  aria-label="Clear search input"
                  class="bx--search-close bx--search-close--hidden"
                  title="Clear input"
                  type="button"
                >
                  <svg
                    aria-label="Clear search input"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    style="will-change: transform;"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12 4.7l-.7-.7L8 7.3 4.7 4l-.7.7L7.3 8 4 11.3l.7.7L8 8.7l3.3 3.3.7-.7L8.7 8z"
                    />
                  </svg>
                </button>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
